app/src/main/java/com/example/reduxtestapp/redux/middleware/CountryMiddleware.kt

```diff
@@ -25,59 +25,66 @@ class CountryMiddleware (
         )
     }
 
-    private fun handleError(error: ErrorState, store: Store<AppState>) {
-        when (error) {
+    private fun handleError(error: ErrorState): MiddlewareResult {
+        return when (error) {
             is ErrorState.EmptyListError -> {
-                store.dispatch(Action.Country.UpdateCountryList(listOf()))
+                MiddlewareResult.ResultAction(Action.Country.UpdateCountryList(listOf()))
             }
             else -> {
                 logError("countryMiddleware: error occurred $error")
-                store.dispatch(Action.Country.Error())
+                MiddlewareResult.ResultAction(Action.Country.Error())
             }
         }
     }
 
-    override fun middleware(store: Store<AppState>, action: Any) {
-        when (action) {
+    override fun handleAction(state: AppState, action: Any): MiddlewareResult? {
+        return when (action) {
             is Action.Country.GetCountries -> {
-                store.dispatch(Action.Async{
+                MiddlewareResult.Async{
                     val networkResult = getCountriesUseCase()
+                    var result: MiddlewareResult? = null
                     networkResult
                         .onRight { countryList ->
-                            store.dispatch(Action.Country.UpdateCountryList(countryList))
+                            result = MiddlewareResult.ResultAction(Action.Country.UpdateCountryList(countryList))
                         }
                         .onLeft { error ->
-                            handleError(error, store)
+                            result = handleError(error)
                         }
-                })
+                    result!!
```

Comment:
check Either's `fold` if you want to extract the value out of it. In general, Either provides many tools so you don't have to do something like this manually - check `map`, `flatMap` and `getOrElse`